List existing cloud connections from all instances.

kishen-v opened the pull request:
Fixes: #518 

Changes in this PR:
1. List all cloud connections that are existent in an account.
```
Kishens-MacBook-Pro bin : list-cc :ᐅ  ./pvsadm get cloud-connections --instance-id 2313fc87-f7e9-4df6-969e-c0d37b8818a3
I0112 19:14:42.567381   91239 root.go:50] Using an API key from IBMCLOUD_API_KEY environment variable
I0112 19:14:48.005075   91239 cloudconnections.go:50] Listing cloud connections across all services, please wait..
+--------------------------------------+------------------------------------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|        CLOUD CONNECTION ID           |                   NAME                   |   STATE     |                                                                             SERVICE NAME                                                                              |
+--------------------------------------+------------------------------------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 3984e503-0333-4ac9-ad48-55c2cd28f3c2 | hypershift-clusterbot-ci-lon-cc          | established | hypershift-clusterbot-ci-lon-pvs                                                                                                                                      |
| f072a4fb-cdfd-4cac-ba06-0c6dc79d7b65 | rh-upstream-hypershift-powervs-ci-cc     | established | rh-upstream-hypershift-powervs-ci-pvs,rh-upstream-hypershift-powervs-ci-e2e-pvs                                                                                       |
| aa50ffa5-1d38-4f30-8fb1-26e1c9023b12 | rh-upstream-hypershift-powervs-ci-e2e-cc | established | rh-upstream-hypershift-powervs-ci-pvs,rh-upstream-hypershift-powervs-ci-e2e-pvs                                                                                       |
| c0b71835-bb43-49b8-af78-0fddaac5cf95 | rh-upstream-hypershift-agent-ci-cc       | established | rh-upstream-hypershift-agent-ci,hypershift-ci-20230518075033-infra-pvs,keerthanaworkspace,test-hyp-pvs,hyp-dhar-tok-4-nodepool,rh-upstream-hypershift-cluster-bot-pvs |
| 332becce-9c97-43c5-adf6-affdae46256d | rh-upstream-hypershift-cluster-bot-cc    | established | test-hyp-pvs,hyp-dhar-tok-4-nodepool,rh-upstream-hypershift-cluster-bot-pvs,rh-upstream-hypershift-agent-ci,hypershift-ci-20230518075033-infra-pvs,keerthanaworkspace |
+--------------------------------------+------------------------------------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Karthik-K-N reviewed (commented):
Couple of queries,
1. Is instance-id needed? can't we list cloud connection in account.
2. Lets also print under which region and zone the particular cloud connection exist.

kishen-v commented:
@Karthik-K-N, 
1. I'll check through the SDKs and find out if there's a account level listing of cloud connections available.
2. Noted.. I'll try get the information related to zones/regions as well.

mkumatag commented:
@kishen-v @Karthik-K-N lets review and merge this PR

kishen-v commented:
A few changes since the last review:
1. No dependency on service ID, the command works as is, as long as the IBMCLOUD_API_KEY is provided.
2. The SDKs do not seem to support region/zone of the cloud connection although the UI seems to display information related to it. 

Thanks!

Karthik-K-N reviewed (commented):
I think overall it can simplified, we can avoid creating client each time on loop as well.

Use resourcecontrollerv2 to list workspaces with gives list of all workspaces along with id and their zone, then for each id create only cloud connection client. while printing print zone along with workspace name.

kishen-v commented on cmd/get/cloudconnections/cloudconnections.go:
```diff
@@ -0,0 +1,97 @@
+var Cmd = &cobra.Command{
+	Use:   "cloud-connections",
+	Short: "List the existing cloud connections in the account",
+	Long:  "List the existing cloud connections enabled across instances in the account",
```
Noted!

Karthik-K-N reviewed (commented):
Thank you for this
/lgtm
/hold for others review.

mkumatag commented on cmd/get/cloudconnections/cloudconnections.go:
```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 IBM Corp
```
update the year to 2024

mkumatag commented on cmd/get/cloudconnections/cloudconnections.go:
```diff
@@ -0,0 +1,137 @@
+		klog.Info("Listing cloud connections across all workspaces, please wait..")
+		// Create a IBM PI Session per zone and reuse them across the workspaces in the same zone.
+		for workspaceZone, workspaces := range zoneWorkspaces {
+			authenticator := &core.IamAuthenticator{ApiKey: c.Config.BluemixAPIKey, URL: *c.Config.TokenProviderEndpoint}
```
can we move this outside of the for loop to reuse?

kishen-v replied:
Made the required change.
Thanks!

mkumatag commented on pkg/client/cloudconnection/cloudconnection.go:
```diff
@@ -0,0 +1,44 @@
+// Copyright 2021 IBM Corp
```
same as above

mkumatag reviewed (changes requested):
minor commets

mkumatag reviewed (approved):
/lgtm

mkumatag commented:
/unhold
